server/src/main/java/org/elasticsearch/action/admin/indices/shrink/TargetNumberOfShardsCalculator.java

```diff
@@ -0,0 +1,190 @@
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.admin.indices.shrink;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.index.IndexWriter;
+import org.elasticsearch.action.admin.indices.stats.IndexShardStats;
+import org.elasticsearch.action.admin.indices.stats.IndicesStatsResponse;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.index.shard.DocsStats;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.store.StoreStats;
+
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.IntFunction;
+
+/**
+ * This class calculates and verifies the number of shards of a target index after a resize operation. It supports three operations SHRINK,
+ * SPLIT and CLONE.
+ */
+public interface TargetNumberOfShardsCalculator {
+
+    /**
+     * Calculates the target number of shards based on the parameters of the request
+     * @param numberOfShards requested number of shards or null if it was not provided
+     * @param maxPrimaryShardSize requested max primary shard size or null if it was not provided
+     * @param sourceMetadata the index metadata of the source index
+     * @return the number of shards for the target index
+     */
+    int calculate(@Nullable Integer numberOfShards, ByteSizeValue maxPrimaryShardSize, IndexMetadata sourceMetadata);
+
+    /**
+     * Verifies if the target number of shards is feasible based on the operation. For example, in the case of SHRINK it will check if the
+     * doc count per shard is within limits and in the other opetations it will ensure we get the right exceptions if the number of shards
+     * is wrong or less than etc.
+     * @param numberOfShards the number of shards the target index is going to have
+     * @param sourceMetadata the index metadata of the source index
+     */
+    void verify(int numberOfShards, IndexMetadata sourceMetadata);
+
+    class Shrink implements TargetNumberOfShardsCalculator {
+        private static final Logger logger = LogManager.getLogger(TransportResizeAction.class);
+
+        private final StoreStats indexStoreStats;
+        private final IntFunction<DocsStats> perShardDocStats;
+
+        public Shrink(String sourceIndex, IndicesStatsResponse indicesStatsResponse) {
+            this.indexStoreStats = indicesStatsResponse.getPrimaries().store;
+            this.perShardDocStats = i -> {
+                IndexShardStats shard = indicesStatsResponse.getIndex(sourceIndex).getIndexShards().get(i);
+                return shard == null ? null : shard.getPrimary().getDocs();
+            };
+        }
+
+        // Facilitates testing
+        public Shrink(StoreStats indexStoreStats, IntFunction<DocsStats> perShardDocStats) {
```

Comment:
I wanted to avoid defining the `perShardDocStats` in every call, but maybe it's not that bad and I like having only one constructor for this. I will change it.